Neos.ContentRepository.Core/Classes/CommandHandler/CommandSimulator.php

```diff
@@ -0,0 +1,167 @@
+ * projection updates) (via {@see CommandSimulator::run()}).
+ * - Create an InMemoryEventStore which buffers created events by command handlers.
+ * - execute all commands via {@see CommandSimulator::handle()}
+ * - -> this will do all constraint checks based on the projection in the open transaction (so it sees
+ * previously modified projection state which is not committed)
+ * - -> it will run the command handlers, buffer all emitted events in the InMemoryEventStore
+ *   -> note to avoid full recursion the workspace command handler is not included in the bus
+ * - -> update the GraphProjection, but WITHOUT committing the transaction.
+ *
+ * This is quite performant because we do not need to fork a new content stream.
+ *
+ * @internal
+ */
+final class CommandSimulator
+{
+    private CommandsThatFailedDuringRebase $commandsThatFailedDuringRebase;
+
+    public function __construct(
+        private readonly ContentGraphProjectionInterface $contentRepositoryProjection,
+        private readonly EventNormalizer $eventNormalizer,
+        private readonly CommandBus $commandBus,
+        private readonly InMemoryEventStore $inMemoryEventStore,
+        private readonly WorkspaceName $workspaceNameToSimulateIn,
+    ) {
+        $this->commandsThatFailedDuringRebase = new CommandsThatFailedDuringRebase();
+    }
+
+    /**
+     * @template T
+     * @param callable(callable(RebaseableCommand): void): T $fn
+     * @return T
+     */
+    public function run(callable $fn): mixed
+    {
+        return $this->contentRepositoryProjection->inSimulation(fn () => $fn($this->handle(...)));
+    }
+
+    /**
+     * Handle a command within a running simulation, otherwise throw.
+     *
+     * We will automatically copy given commands to the workspace this simulation
+     * is running in to ensure consistency in the simulations constraint checks.
+     */
+    private function handle(RebaseableCommand $rebaseableCommand): void
+    {
+        // FIXME: Check if workspace already matches and skip this, e.g. $commandInWorkspace = $command->getWorkspaceName()->equals($this->workspaceNameToSimulateIn) ? $command : $command->createCopyForWorkspace($this->workspaceNameToSimulateIn);
+        // when https://github.com/neos/neos-development-collection/pull/5298 is merged
+        $commandInWorkspace = $rebaseableCommand->originalCommand->createCopyForWorkspace($this->workspaceNameToSimulateIn);
+
+        try {
+            $eventsToPublish = $this->commandBus->handle($commandInWorkspace);
+        } catch (\Exception $exception) {
+            $this->commandsThatFailedDuringRebase = $this->commandsThatFailedDuringRebase->withAppended(
+                new CommandThatFailedDuringRebase(
+                    $rebaseableCommand->originalSequenceNumber,
+                    $rebaseableCommand->originalCommand,
+                    $exception
+                )
+            );
+
+            return;
+        }
```

Comment:
as this got really repetitive over the 5 usages i added the logic to collect the failures to the command simulator.

```php
$commandSimulator = $this->commandSimulatorFactory->createSimulator($baseWorkspace->workspaceName);

$commandSimulator->run(
    static function ($handle) use ($rebaseableCommands): void {
        foreach ($rebaseableCommands as $rebaseableCommand) {
            $handle($rebaseableCommand);
        }
    }
);

if ($commandSimulator->hasCommandsThatFailed()) {
    throw WorkspaceRebaseFailed::duringPublish($commandSimulator->getCommandsThatFailed());
}
```

i thought of returning it first but that would mean that the return value of `$handle` has to be manually collected or we create a lot of logic in the `run` closure ... as we do the same with `$commandSimulator->eventStream()` i think its fine to do:) The object has state already ^^

Reply:
Does it have state? I thought by putting handle as closure we got rid of all the state? But anyways looks good.

Reply:
not _itself_ ... it was `readonly` but as the class contains and exposes the in-memory event store this is really stateful already ;)